File: Data Structures & Algorithms/two-integer-sum/submission-3.java
class Solution {
    public int[] twoSum(int[] nums, int target) {
     Map<Integer, Integer> indexOf = new HashMap<>();
     for (int i = 0; i < nums.length; i++) {
        int x = nums[i];
        int complement = target - x;
        if (indexOf.containsKey(complement)) {
            return new int[] { indexOf.get(complement), i};
        }
        indexOf.put(x,i);
     }
     throw new IllegalArgumentException("No two sum solution exists for the given input.");
    }
}

File: Data Structures & Algorithms/target-sum/submission-3.py
class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        S = sum(nums)
        if (S + target) % 2 != 0:
            return 0
        P = (S + target) // 2
        if P < 0:
            return 0
        dp = [0] * (P + 1)
        dp[0] = 1
        for num in nums:
            for s in range(P, num - 1, -1):
               dp[s] += dp[s - num]
        return dp[P]    
        